File: android/app/src/main/java/com/ventouxlabs/gatepath/ui/PortalLoadError.kt
package com.ventouxlabs.gatepath.ui

/**
 * Pure, JVM-testable model for "the portal page did not load".
 *
 * Every load failure in [GatepathWebView] previously rendered identically: a
 * blank white page, with the only trace a `Log.w` line the user can't see.
 * That covers cert errors, DNS failures, EPERM from the sandboxed WebView
 * process, redirect loops — all of them. This type turns a failure into
 * something the UI can show.
 *
 * Deliberately free of `android.*` imports so the classification and copy can
 * be regression-tested on a plain JVM (same reason as [WebViewHostMatching]
 * and [SslErrorPolicy]). The WebView error constants are duplicated as
 * literals below rather than imported — see [PortalLoadErrorKind.fromWebViewErrorCode].
 */
data class PortalLoadError(
    val kind: PortalLoadErrorKind,
    /**
     * Host of the failed URL, or blank if it couldn't be parsed. Host only,
     * never the full URL: portal URLs routinely carry MAC addresses, gateway
     * IPs and session tokens in query params (same reasoning as the
     * release-build logging policy in [GatepathWebView]).
     */
    val host: String,
    /** Raw code/description for the log line. Never shown to the user. */
    val technicalDetail: String,
)

enum class PortalLoadErrorKind {
    /** Certificate error on a host outside the portal's origin — refused by [SslErrorPolicy]. */
    CERT_REJECTED,

    /** Name resolution failed. Common when the portal hands out a hostname only it can resolve. */
    HOST_LOOKUP_FAILED,

    /** TCP connect / IO failure or timeout. */
    UNREACHABLE,

    /** The gateway bounced the request around until the WebView gave up. */
    REDIRECT_LOOP,

    /** TLS handshake failed outright (not a cert-trust decision — no page to proceed to). */
    TLS_HANDSHAKE_FAILED,

    /** Anything else, including the WebView's own ERROR_UNKNOWN. */
    UNKNOWN,
    ;

    companion object {
        /**
         * Map a `WebViewClient.ERROR_*` code to a kind.
         *
         * The constants are inlined as literals so this file stays
         * `android.*`-free and testable on the JVM. They are stable public API
         * (`android.webkit.WebViewClient`) and have not changed since API 1:
         *  -2 ERROR_HOST_LOOKUP, -4 ERROR_AUTHENTICATION, -6 ERROR_CONNECT,
         *  -7 ERROR_IO, -8 ERROR_TIMEOUT, -9 ERROR_REDIRECT_LOOP,
         *  -11 ERROR_FAILED_SSL_HANDSHAKE.
         */
        fun fromWebViewErrorCode(code: Int): PortalLoadErrorKind = when (code) {
            -2 -> HOST_LOOKUP_FAILED
            -6, -7, -8 -> UNREACHABLE
            -9 -> REDIRECT_LOOP
            -11 -> TLS_HANDSHAKE_FAILED
            else -> UNKNOWN
        }
    }
}

/**
 * User-facing copy for a [PortalLoadError].
 *
 * Rules this copy follows:
 *  - Say what failed and what the user can do. "Something went wrong" is what
 *    we're replacing, so it is not an acceptable fallback.
 *  - Never blame the user's device for a gateway problem — captive gateways
 *    are the usual culprit and the wording should point there.
 *  - Host is included only when non-blank, and it is a host, never a full URL.
 */
object PortalLoadErrorText {

    fun title(kind: PortalLoadErrorKind): String = when (kind) {
        PortalLoadErrorKind.CERT_REJECTED -> "Sign-in page blocked for safety"
        PortalLoadErrorKind.HOST_LOOKUP_FAILED -> "Couldn't find the sign-in page"
        PortalLoadErrorKind.UNREACHABLE -> "Couldn't reach the sign-in page"
        PortalLoadErrorKind.REDIRECT_LOOP -> "The network kept redirecting"
        PortalLoadErrorKind.TLS_HANDSHAKE_FAILED -> "Secure connection failed"
        PortalLoadErrorKind.UNKNOWN -> "The sign-in page didn't load"
    }

    fun body(error: PortalLoadError): String {
        val where = if (error.host.isBlank()) "this network's sign-in page" else error.host
        return when (error.kind) {
            PortalLoadErrorKind.CERT_REJECTED ->
                "$where sent an invalid security certificate, and it isn't the " +
                    "network's own sign-in page — so Gatepath refused to load it. " +
                    "This can mean the network is tampering with traffic. " +
                    "Avoid signing in here."

            PortalLoadErrorKind.HOST_LOOKUP_FAILED ->
                "Gatepath couldn't look up $where. The network may still be " +
                    "setting up, or its DNS may be misconfigured. Try again in a moment."

            PortalLoadErrorKind.UNREACHABLE ->
                "Gatepath reached the network but $where didn't respond. " +
                    "The gateway may be busy or offline. Try again."

            PortalLoadErrorKind.REDIRECT_LOOP ->
                "$where redirected in a loop and never settled on a page. " +
                    "This is usually a fault in the network's sign-in system. " +
                    "Try again, or use the network's own sign-in screen."

            PortalLoadErrorKind.TLS_HANDSHAKE_FAILED ->
                "Gatepath couldn't establish a secure connection to $where. " +
                    "The gateway may not support a compatible encryption standard."

            PortalLoadErrorKind.UNKNOWN ->
                "Gatepath couldn't load $where and the network didn't say why. " +
                    "Try again, or run diagnostics to look closer."
        }
    }

    /** Whether a retry has any chance of helping. Drives the Try-again button. */
    fun isRetryable(kind: PortalLoadErrorKind): Boolean = when (kind) {
        // Retrying a rejected certificate just re-rejects it, and inviting the
        // user to retry a possible tampering signal is the wrong nudge.
        PortalLoadErrorKind.CERT_REJECTED -> false
        else -> true
    }
}

File: desktop/gatepath/portal_load_error.py
"""Pure model for "the desktop portal page did not load".

No `gi` / WebKit imports — same headless-safe contract as
`gatepath.ui.diagnosis_panel`'s pure layer, so the classification and the
user-facing copy are unit-testable without a GTK environment.

Counterpart to Android's `ui/PortalLoadError.kt`. The two apps share no code by
design, so this is a deliberate re-implementation, not an import.
"""

from __future__ import annotations

import enum
from dataclasses import dataclass

# WebKit GError domains, as the strings PyGObject exposes on `GLib.Error.domain`.
DOMAIN_NETWORK = "WebKitNetworkError"
DOMAIN_POLICY = "WebKitPolicyError"

# WebKitNetworkError codes.
_NETWORK_FAILED = 199
_NETWORK_TRANSPORT = 300
_NETWORK_UNKNOWN_PROTOCOL = 301
_NETWORK_CANCELLED = 302
_NETWORK_FILE_DOES_NOT_EXIST = 303

# WebKitPolicyError codes.
_POLICY_FRAME_LOAD_INTERRUPTED = 202


class PortalLoadErrorKind(enum.Enum):
    """Why the portal page failed to render."""

    #: Certificate could not be trusted and no bypass applied.
    CERT_REJECTED = "cert_rejected"
    #: Name resolution / transport failure reaching the gateway.
    UNREACHABLE = "unreachable"
    #: The gateway offered something the WebView can't display.
    UNSUPPORTED = "unsupported"
    #: Anything else, including WebKit's own generic failure.
    UNKNOWN = "unknown"


@dataclass(frozen=True)
class PortalLoadError:
    """A load failure worth showing the user.

    `host` is a host, never a full URL: portal URLs routinely carry MAC
    addresses, gateway IPs and session tokens in query params, and this string
    reaches the screen.
    """

    kind: PortalLoadErrorKind
    host: str
    technical_detail: str


def is_deliberate_cancel(domain: str, code: int) -> bool:
    """True when *we* stopped the load, so no error should be shown.

    `_on_decide_policy` calls `decision.ignore()` for off-domain navigation.
    WebKit reports that as a load failure like any other — without this guard,
    every deliberately-refused navigation would pop an error panel at the user
    as though something had gone wrong.
    """
    if domain == DOMAIN_POLICY and code == _POLICY_FRAME_LOAD_INTERRUPTED:
        return True
    return domain == DOMAIN_NETWORK and code == _NETWORK_CANCELLED


def classify(domain: str, code: int) -> PortalLoadErrorKind:
    """Map a WebKit GError domain/code pair to a kind.

    Unrecognised domains and codes fall back to UNKNOWN rather than raising —
    a surprising error code must still produce a visible explanation, since
    the alternative is the silent failure this module exists to prevent.
    """
    if domain == DOMAIN_NETWORK:
        if code in (_NETWORK_TRANSPORT, _NETWORK_FAILED):
            return PortalLoadErrorKind.UNREACHABLE
        if code in (_NETWORK_UNKNOWN_PROTOCOL, _NETWORK_FILE_DOES_NOT_EXIST):
            return PortalLoadErrorKind.UNSUPPORTED
    if domain == DOMAIN_POLICY:
        return PortalLoadErrorKind.UNSUPPORTED
    return PortalLoadErrorKind.UNKNOWN


_TITLES = {
    PortalLoadErrorKind.CERT_REJECTED: "Sign-in page blocked for safety",
    PortalLoadErrorKind.UNREACHABLE: "Couldn't reach the sign-in page",
    PortalLoadErrorKind.UNSUPPORTED: "The sign-in page couldn't be displayed",
    PortalLoadErrorKind.UNKNOWN: "The sign-in page didn't load",
}


def title(kind: PortalLoadErrorKind) -> str:
    """Short headline for *kind*. Never empty."""
    return _TITLES[kind]


def body(error: PortalLoadError) -> str:
    """Explanation for *error*, naming the host when one is known.

    Points at the gateway rather than the user's machine — a captive gateway is
    the usual culprit and the copy shouldn't send people to debug their laptop.
    """
    where = error.host.strip() or "this network's sign-in page"
    if error.kind is PortalLoadErrorKind.CERT_REJECTED:
        return (
            f"{where} sent an invalid security certificate, so Gatepath "
            "refused to load it. This can mean the network is tampering with "
            "traffic. Avoid signing in here."
        )
    if error.kind is PortalLoadErrorKind.UNREACHABLE:
        return (
            f"Gatepath couldn't reach {where}. The network may still be "
            "setting up, or the gateway may be offline. Try again in a moment."
        )
    if error.kind is PortalLoadErrorKind.UNSUPPORTED:
        return (
            f"{where} returned something Gatepath can't display. This is "
            "usually a fault in the network's sign-in system."
        )
    return (
        f"Gatepath couldn't load {where} and the network didn't say why. "
        "Try again, or run diagnostics to look closer."
    )


def is_retryable(kind: PortalLoadErrorKind) -> bool:
    """Whether to offer a retry.

    Certificate rejections are not retryable: retrying re-rejects, and nudging
    the user past a possible tampering signal is the wrong affordance.
    """
    return kind is not PortalLoadErrorKind.CERT_REJECTED
